### ISSUE TITLE
Discrepancy Between CoolProp DLL and Online Calculator Outputs

### ISSUE THREAD

**Issue by gehadrabee24** [CoolProp/CoolProp#2434]
**Description**
I am encountering discrepancies between the CoolProp DLL outputs used in TRNSYS and the results provided by the CoolProp online calculator. Specifically, there are differences in calculated enthalpy values and errors under certain conditions when using the DLL in TRNSYS.

**Steps to Reproduce**
Pressure and Temperature Discrepancy:

Set temperature to 27°C.
Set pressure to 9.9 MPa.
Run enthalpy calculation using both the CoolProp DLL in TRNSYS and the CoolProp online calculator.
Expected behavior: Enthalpy output should be consistent across both the DLL and the online calculator.
Actual behavior: The DLL returns an enthalpy value of 426 kJ/kg, whereas the online calculator returns 406 kJ/kg for the same input conditions.

Error on Specific Enthalpy and Pressure Inputs:

Set enthalpy to 104.040 kJ/kg.
Set pressure to 9.998999 MPa.
Run temperature calculation using both the CoolProp DLL in TRNSYS and the online calculator.
Expected behavior: Both the DLL and the online calculator should return a valid temperature, approximately 128.08 K.
Actual behavior: The DLL outputs an “invalid temperature” error, while the online calculator provides a temperature of 128.08 K.

Versions
CoolProp Version: Version used in TRNSYS 18
Operating System and Version: Windows OS
Access Method: TRNSYS integration with CoolProp DLL

Additional Information
Any guidance on addressing this discrepancy or suggested troubleshooting steps would be greatly appreciated. If possible, could you test the enthalpy output for 27°C and 9.9 MPa to verify if this discrepancy persists outside of TRNSYS?

Thank you for your assistance!

**Comment by gehadrabee24** [comment]
**Additional Information**  
For the second issue (temperature calculation at enthalpy = 104.040 kJ/kg and pressure = 9.998999 MPa), I initially reached out to the TRNSYS support team. They recommended consulting the CoolProp team directly regarding the “invalid floating point” error encountered in the DLL integration. Any guidance on addressing this discrepancy or troubleshooting steps would be greatly appreciated. 

Thank you for your assistance!

**Comment by ibell** [comment]
As always the question is what reference states are being used in each source. If they are not the same, you can't use enthalpies from one and put it into the other. So first thing is to make sure the reference states are the same.

**Comment by gehadrabee24** [comment]
I’m using the CoolProp DLL integrated with TRNSYS 18.06, but I couldn’t find clarification in the TRNSYS documentation on the specific reference state used in the DLL. 

Notably, at higher enthalpy and pressure values, the CoolProp DLL in TRNSYS matches the values from the CoolProp online calculator, which suggests that the same reference state might be used. However, discrepancies appear at lower values, as described in the initial post. 

Thank you for your assistance!

**Comment by ibell** [comment]
1. Without knowing what version of CoolProp is being used it is impossible to say anything. 
2. Can you please show your calculated values that are the same as TRNSYS? And confirm you get the same with CoolProp 6.6.0? What fluid is this for?

**Comment by gehadrabee24** [comment]
It's not mentioned in TRNSYS what is the version, I will ask the Technical Support, but here is the values: If I run with T = -145.15 C and enthalpy = 104.04 kJ/kg CoolProp returns with an error. which is not the same at the online cool prop, and  If I run with T = -145.15 C and P = 9.998999 MPa I get an enthalpy of 103.866 kJ/kg

**Comment by ibell** [comment]
Yeah TH are not supported in CoolProp due to the possibility of multiple solutions.  

I did:
```python
import CoolProp.CoolProp as CP
print(CP.PropsSI('Hmass', 'T', -145.15+273.15, 'P', 10e6, 'Air'))
 ```
which yields
```
103856.23303619286
```

**Comment by gehadrabee24** [comment]
So I cannot get the temperature from the enthalpy and the pressure?

**Comment by ibell** [comment]
Sure you can: 
```python
import CoolProp.CoolProp as CP
h = CP.PropsSI('Hmass', 'T', -145.15+273.15, 'P', 10e6, 'Air')
T_hp = CP.PropsSI('T', 'Hmass', h, 'P', 10e6, 'Air')
print(h, T_hp)
```
yielding
```
103856.23303619286 128.00000000000423
```

**Comment by gehadrabee24** [comment]
Okay, what Im exactly trying to do is to call the temperature by the enthalpy and pressure
nRef = 729 ! Air
        prop(3) = 104.040 ! Enthalpy in kJ/kg
        prop(2) = 9.998999 ! Pressure in MPa
        Call CoolProp_Fluid_Properties(prop, nRef, 23, iFlagr)
        If (ErrorFound()) Return
        Temperature = prop(1)
 And an error is generated, while it is not the same on the online coolprop, can’t I just call the temperature with the enthalpy and pressure only?

**Comment by ibell** [comment]
Sure you can. That's what I did in the code above.

**Comment by gehadrabee24** [comment]
It returns an error, and the technical support of trnsys asked me to contact you as it is an error with the coolprop dll file

**Comment by ibell** [comment]
To reproduce: you need to calculate H from T,P, then put THAT h,p in. That will eliminate the reference state question. You must be quite close to the critical point? That is always a problem.

**Comment by ibell** [comment]
Also, what precise error message do you obtain from TRNSYS?

**Comment by gehadrabee24** [comment]
No, I get the enthalpy from other calculations and also the pressure, then I need to get the temperature of this state, the error generated is invalid floating point

**Comment by ibell** [comment]
Hmm. How does TRNSYS call CoolProp? Your input gives a reasonable value in Python:
```python
import CoolProp.CoolProp as CP
T_hp = CP.PropsSI('T', 'Hmass', 104.040e3, 'P', 10e6, 'Air')
print(T_hp)
```
yielding
```
128.0827784501017
```

**Comment by gehadrabee24** [comment]
The calling function i sent in the above code, it calls the cool prop dll, it is a file inside trnsys library as i know

**Comment by ibell** [comment]
There must be at least one more interface layer because that is not a CoolProp function you call

**Comment by gehadrabee24** [comment]
Also there is interface file called coolprop interface

**Comment by ibell** [comment]
Yes, that interface. Do you get the same error if you ONLY call that function and do nothing else? I strongly believe this is TRNSYS related, and not having anything to do with CoolProp. 

One thing that it could be related to... is the FPU error bits. When you do calculations on your computer, the processor keeps track of the FPU (floating point unit) error flags. CoolProp uses NAN values as placeholders which will set FPU error bits, but the calls _within_ the function reset the flags to avoid similar errors I have seen in other environments (I think Delphi is the one I remember). Resetting the FPU error bits was the only way to get around it: https://github.com/CoolProp/CoolProp/blob/master/src/CoolPropLib.cpp#L54. I wonder if TRNSYS does a check and immediately errors out, with some sort of interrupt strategy so that we don't get a chance to reset the FPU bits, causing this issue.

**Comment by gehadrabee24** [comment]
Okay, I have forwarded this explanation to TRNSYS support team and they replied to me by this
![IMG_6840](https://github.com/user-attachments/assets/65af6580-6f0f-4440-b2d9-f7efe3f5feb8)
![IMG_6841](https://github.com/user-attachments/assets/eb5124d9-f4df-44d5-979e-e02aa4a4d6eb)

**Comment by gehadrabee24** [comment]
I just can not forward it here
![IMG_6842](https://github.com/user-attachments/assets/de609bfb-755d-4268-8d10-9aabc518c2f4)
![IMG_6843](https://github.com/user-attachments/assets/fde75362-2bd2-4370-83e2-66da8459535c)
![Uploading IMG_6842.png…]()
![IMG_6843](https://github.com/user-attachments/assets/1ad2cffa-f105-4b07-8b83-8971304d8dd3)

**Comment by gehadrabee24** [comment]
Also they mentioned that they are using cool prop version 6.6.0

**Comment by ibell** [comment]
I really don't think this has anything to do with CoolProp. I remain convinced the issue is the FPU bit flags. CoolProp does not return that error code in general, and not for these particular inputs either. I can debug with someone from TRNSYS if they want to dig in. Also, their calling code is horrendously inefficient; it could be 5 times faster at minimum by doing p,h -> T,density, and then all other calls with T,density.

**Comment by ibell** [comment]
Figured out a robust solution in #3016. We capture the FPU bits at entry to any CoolPropLib.h method, ignore any FPU flag bit changes that happen within a CoolProp call, and then restore the flags on exit. This appears to be a FORTRAN "feature" that we have to work around.
